containers-contrib-readonly/src/Map/Contrib/Laws.hs

```diff
@@ -1,8 +1,77 @@
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE ViewPatterns #-}
+
 module Map.Contrib.Laws
        ( singletonFromList
+       , nullEmpty
+       , nullImpliesZeroSize
+       , nonZeroSizeImpliesNotNull
+       , sizeIsNatural
+       , emptyZeroSized
+       , singletonOneSized
+       , memberEmptyFalse
+       , memberSingletonSame
+       , newMemberYieldsValidValue
+       , lookupDefaultEmpty
+       , listToSingleton
+       , keysOfSingleton
+       , elemsOfSingleton
+       , lookupMatchMember
        ) where
 
-import Map (Key, fromList, singleton)
+import Map
+import Prelude hiding (lookup, null)
+import           Test.QuickCheck
+
+nullEmpty :: Bool
+nullEmpty = null empty
+
+nullImpliesZeroSize :: Key k => Gen [(k, v)] -> Gen Property
+nullImpliesZeroSize pairs = do
+  m <- fromList <$> pairs
+  pure $ null m ==> size m == 0
+
+nonZeroSizeImpliesNotNull :: Key k => [(k, v)] -> Property
+nonZeroSizeImpliesNotNull (fromList -> m) = size m > 0 ==> not (null m)
```

Comment:
Are there any differences between passing `Gen [(k, v)]` and just `[(k, v)]` to functions? Also, what do you think about passing just `Map k v` directly and have generators implemented on the call-site?

Reply:
We can't have `Gen a` as parameter because the parameters need to have `Arbitrary` instances, and we don't have the instance `Arbitrary a => Arbitrary (Gen a)` implemented.

For the map side, in order to pass Map caller should know the `Map` type of the called module (so I would need to write `quickCheck $ PR.nullImpliesZeroSize PR.Map.empty`). But with the improvement on having modules implement `checkLaws` themselves, passing `Map k v` instead of pair list is just trivial, so I'll do that :)